pkg/oauth/user.go

```diff
@@ -207,16 +211,49 @@ func (s *server) Login(c echo.Context) error {
 	return c.NoContent(http.StatusNoContent)
 }
 
+var errInvalidLogoutRedirectURI = errors.DefinePermissionDenied("invalid_logout_redirect_uri", "the redirect URI did not match the one(s) defined in the client")
+var errMissingAccessTokenIDParam = errors.DefinePermissionDenied("missing_param_access_token_id", "Access token ID was not provided")
+
 func (s *server) Logout(c echo.Context) error {
 	ctx := c.Request().Context()
-	session, err := s.getSession(c)
+	accessTokenID := c.QueryParam("access_token_id")
+	if accessTokenID == "" {
+		return errMissingAccessTokenIDParam
+	}
+	at, err := s.store.GetAccessToken(ctx, accessTokenID)
 	if err != nil {
 		return err
 	}
-	events.Publish(evtUserLogout(ctx, session.UserIdentifiers, nil))
-	if err = s.store.DeleteSession(ctx, &session.UserIdentifiers, session.SessionID); err != nil {
+	if err = s.store.DeleteAccessToken(ctx, accessTokenID); err != nil {
 		return err
 	}
 	s.removeAuthCookie(c)
-	return c.NoContent(http.StatusNoContent)
+	events.Publish(evtUserLogout(ctx, at.UserIDs, nil))
+	if err = s.store.DeleteSession(ctx, &at.UserIDs, at.UserSessionID); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	}
+	client, err := s.store.GetClient(ctx, &at.ClientIDs, &types.FieldMask{Paths: []string{"logout_redirect_uris"}})
+	if err != nil {
+		return err
+	}
+	var redirectURI = ""
+	redirectParam := c.QueryParam("post_logout_redirect_uri")
+	if redirectParam == "" {
+		redirectURI = client.LogoutRedirectURIs[0]
+		if redirectURI == "" {
+			redirectURI = s.config.UI.MountPath()
+		}
+	} else {
+		redirectURI, err = osin.ValidateUriList(strings.Join(client.LogoutRedirectURIs, ","), redirectParam, ",")
```

Comment:
Better to use `osin.ValidateUri` instead of joining, and then letting `osin.ValidateUriList` split them again.